### PR TITLE
python: introduce new service for `flux account` commands

### PR DESCRIPTION
**Pull request by cmoussa1**
#### Background

`flux account` commands can currently only be executed on the management node of a cluster (where the flux-accounting DB is installed). So, users/admins who want to run simple `flux account` commands to see user or bank info cannot run these commands from other nodes on a cluster. There is a need to improve this and enable commands to interface with the DB "remotely."

---

This is an early [WIP] PR that introduces a new service to flux-accounting that can be managed by `systemd` and is designed to run on the rank 0 broker of a Flux system instance. This service, `flux-account-service.py`, is installed to `${prefix}/libexec/flux/cmd` and can be started _after_ the creation of a flux-accounting database:

```console
$ flux account-service -p path/to/db &
[1] 1234
```

If no path is specified on the command line when the service is started, it will try to connect to the database in its default location, `/var/lib/flux/FluxAccounting.db`. The service is now the component responsible for maintaining a connection to the SQLite database instead of the front-end script that accepts `flux account` commands, `flux-account.py`.

`flux-account.py` now issues RPCs to this service instead of calling the Python bindings directly. The service will accept the RPCs from the front-end script, call the appropriate Python binding, and send a payload back to the front-end script with any returned information, where it is unpacked, and if necessary, printed to `STDOUT` or `STDERR`.

These changes did not require any major rework to the existing suite of tests for flux-accounting, other than now starting this new service in the tests where `flux account` commands are called. You'll also notice that the optional path that had to previously be specified when issuing a `flux account` command is no longer needed, since the service is now responsible for connecting to the flux-accounting database. This is true for almost every `flux account` command _except_ for the command that creates the database for the first time, `flux account create-db`, which I have left to directly call the `create_db()` Python function.

I'm leaving this as [WIP] for now until I can do a little cleanup here and there with the commits, and to also get some early feedback on the service, as this is my first time writing something like this, and I'm sure there are improvements that I missed 😄 . I'll list them below:

- [x] check commit messages and contents for misspellings and type of commit
- [x] consult on various `sleep 1` inserts in tests where service is loaded and `flux account` commands are called immediately after
- [x] consult behavior of service when connection to DB fails (because of bad path, some other error occurs, etc.)

### CODE REVIEW

**Comment by cmoussa1** [comment]
OK, I think before I take this out of [WIP] and mark it as ready for review, I should at least consult on a couple of small things regarding testing and some behavior for the service on connection attempts:

1) You'll notice that in some of the sharness test files (specifically, the ones that load the service and then immediately call `flux account` commands in the following test), I've inserted some `sleep 1` commands. This is because I _think_ the service needs another second or two to finish loading before the commands can be called, otherwise the following tests will fail with an error message complaining that the service hasn't been registered yet. I've only had to use `sleep 1` commands, so I don't think it's _that_ intrusive to the tests running in Docker or in GitHub actions, but is there perhaps a better way to this that I am not aware of? I'm thinking of something along the lines of:

```
test_expect_success 'check that service is loaded' '
    <command to check that flux-account-service.py is loaded>
'
```

2) In the code for `flux-account-service.py`, you'll notice that in the method where the service attempts to establish a SQLite connection to the flux-accounting database, in case of a failure (due to a bad path, the DB not existing, or some other error), I call `sys.exit()`, which is what I used to call in the front-end script. Is that the best choice for a service such as this? Would it be better to exit some other way? As proposed in the commit that adds the service, the connection attempt to the SQLite database happens _before_ `handle.reactor_run()` is called, but after a Flux handle is created with `flux.Flux()`. Is calling `sys.exit()` on a failed connection attempt the best approach? Perhaps there is a better way to safely exit/stop the service that I am unaware of.

**Comment by grondo** [comment]
> You'll notice that in some of the sharness test files (specifically, the ones that load the service and then immediately call flux account commands in the following test), I've inserted some sleep 1 commands.

The `sleep 1` isn't a good idea because on a busy system even 1 second may not be enough time, so there could be spurious failures, and extending the sleep isn't great because then you extend the test runtime unnecessarily.

I think the service is guaranteed to be registered after `handle.service_register (...).get()`, i.e. after you get a successful response. You just need some way in testing to synchronize at that point.

The first method that occurs to me is to add a `--daemonize` option to `flux account-service` which would turn the service into a daemon process. The original child could exit once the service is registered and the database is open, in order to return errors to the caller, but after both of these are successful, the process would go into the background and the test could proceed with a guarantee that the service is available.

If you go this route, I'd also add a `shutdown` method to the accounting service which would tell the script to exit.

To daemonize your process you could start with this code (you don't need the pid file and lockfile stuff since in practice systemd will handle that for you)

  https://github.com/thesharp/daemonize/blob/master/daemonize.py#L103


> I call sys.exit(), which is what I used to call in the front-end script. Is that the best choice for a service such as this? Would it be better to exit some other way? 


If the code architecture allows it, I would first try to open the database file and exit with an error to stderr and a well known exit code. When running under systemd, the error will be logged and the special exit code can tell systemd not to restart the service.

If you use the daemonize suggestion above for testing, you'll want to open the DB before daemoninzing so that tthe exit code can get reported to the caller.

Does that make any sense?

**Comment by cmoussa1** [comment]
Thanks for the feedback and suggestions @grondo. Let me take a crack at this and see if I get anywhere with it. I'll report back with questions if they come up!

**Comment by grondo** [comment]
> To daemonize your process you could start with this code

I guess if this is for testing, you can actually start with a very simple function that calls `os.fork()` once and exits the parent. In testing we actually want output from the server process to go to the same stdout/err as the test, so maybe start with:

```
def background():
        pid = os.fork() 
        if pid > 0:
            # exit first parent
            sys.exit(0) 
```

That might get you started. Maybe call this option `--test-background` to avoid confusion with true daemonization.

Then you'll have to update uses of `kill $service_pid` to issue a shutdown request, since `service_pid=$1` will be referencing the wrong process id.

**Comment by cmoussa1** [comment]
Oh interesting. Yeah, I believe the use case I am looking for here is pretty much exclusively for testing and just to avoid having to call `sleep`, which feels not great, lol.

Just as a quick and dirty experiment, I removed the `sleep 1` command from some of my sharness tests and replaced it with a couple of tests: one that creates a small Python script that calls `handle.service_register(...).get()`, since you mentioned that the service is guaranteed to be registered after that's called, and another one that tested that the response received was `None`. Would this also be an okay solution?

**Comment by cmoussa1** [comment]
OK, I _think_ I've got the test option working for the new service so that no `sleep` commands have to be issued anywhere in the sharness tests, as well as not needing to `kill` the processes. As a result, I'll take this out of [WIP] and mark it as ready for review whenever convenient.

**Review by jameshcorbett — APPROVED** [review]
This looks reasonable to me and the implementation makes sense! Just thought I would take a look at what's going on in this repo.

File: src/cmd/flux-account-service.py
Comment (cmoussa1) [reply src/cmd/flux-account-service.py:193]: Hmm, sorry, I might need more clarification here. Are you referring to the two blank lines in between functions? If so, I _think_ those are how the `black` formatter formats Python files for this repo.

Comment (jameshcorbett) [reply src/cmd/flux-account-service.py:193]: I meant you always have a blank line before a `return` statement. I haven't seen it a lot, that's all.

**Review by grondo — CHANGES_REQUESTED** [review]
Ok, I've added some initial comments, mainly on the account-service process.

Should we try to include a systemd unit file before merging this as well? Otherwise as I noted in the first commit message, this is more general than "introduce a new systemd-managed service" (since there's no systemd unit file provided).

File: src/cmd/flux-account-service.py
Comment (cmoussa1) [reply src/cmd/flux-account-service.py:223]: No, I don't think it's necessary. I think I just copied some suggested code in one of the examples in #186 and stuck with it in the implementation, sorry! I can probably just get rid of the `unpack_*` functions. Will do this

**Comment by cmoussa1** [comment]
Thanks for starting to review this @grondo! I've just force-pushed some fixes based on the feedback you gave. Notably, I've:

- edited the commit message to make it more general
- renamed the service to `AccountingService` instead of just `Accounting`
- removed the functions that unpack the `msg` and condensed all of the `watcher_create()`s into a `for` loop
- moved the 
- used `logging.error()` instead of printing any error messages and refined the `FileExistsError` message to more clearly express the error message (`"flux-accounting service is already registered"`)
- moved the registration of the service before going into the background

> Should we try to include a systemd unit file before merging this as well? Otherwise as I noted in the first commit message, this is more general than "introduce a new systemd-managed service" (since there's no systemd unit file provided).

That sounds like a good idea! I'm pretty unfamiliar with systemd unit files though. Could you tell me what that would look like? Or maybe there is already an example in flux-core that I could try to follow?

**Comment by grondo** [comment]
A couple general comments while I give the PR a review:

 * [RFC 1](https://flux-framework.readthedocs.io/projects/flux-rfc/en/latest/spec_1.html#patch-requirements) (C4.1 Patch Requirements) specifies that commit messages should strive to keep commit message title <= 50 characters and wrap the commit message body at 72. At least the commit message bodies in this series seem to go beyond that, so they line wrap when I'm viewing them with `git log`. Not required, but I'd suggest fixing that.
 * RFC 1 also suggests that `Problem:` be used in a commit message body to describe the one problem being solved by the commit and its solution (`Solution:` is not necessary). Again this is not a requirement, but it can be helpful to get into that practice. (e.g. the last commit `.pylintrc: disable some pylint checks` doesn't have a commit message at all, which can be a bit of a red flag - what checks are disabled (very difficult to tell since they're all on one line) and why is it necessary instead of fixing the issue pylint is calling out, or by using a comment to disable as an exception?)
 * My opinion is that all the "change *-foo commands" commits could be squashed into one commit `flux-account.py: update commands to use RPC to new accounting service`, however I leave that up to you.

**Comment by cmoussa1** [comment]
Just force-pushed some fixes based on your feedback. Thanks!

> [RFC 1](https://flux-framework.readthedocs.io/projects/flux-rfc/en/latest/spec_1.html#patch-requirements) (C4.1 Patch Requirements) specifies that commit messages should strive to keep commit message title <= 50 characters and wrap the commit message body at 72. At least the commit message bodies in this series seem to go beyond that, so they line wrap when I'm viewing them with git log. Not required, but I'd suggest fixing that.

Oops, this is totally my fault, sorry about that! For some reason, I thought the character limit for commit message bodies was 80 characters, not 72. I changed these to keep the message bodies at 72 characters or less (let me know if I missed any).

> RFC 1 also suggests that Problem: be used in a commit message body to describe the one problem being solved by the commit and its solution (Solution: is not necessary). Again this is not a requirement, but it can be helpful to get into that practice. (e.g. the last commit .pylintrc: disable some pylint checks doesn't have a commit message at all, which can be a bit of a red flag - what checks are disabled (very difficult to tell since they're all on one line) and why is it necessary instead of fixing the issue pylint is calling out, or by using a comment to disable as an exception?)

Yeah, I should have added a message body to this commit (and the others that didn't have message bodies with `Problem` statements). Turns out I didn't need a couple of the checks I disabled, so I refactored that commit slightly and also added brief explanations to the checks I _did_ end up disabling. It should be viewable in the commit message.

> My opinion is that all the "change *-foo commands" commits could be squashed into one commit flux-account.py: update commands to use RPC to new accounting service, however I leave that up to you.

Sounds good. I just went ahead and squashed all of those into just commit. Much cleaner this way IMHO. Thanks for the suggestion.

**Comment by grondo** [comment]
> Turns out I didn't need a couple of the checks I disabled, so I refactored that commit slightly and also added brief explanations to the checks I did end up disabling. It should be viewable in the commit message.

Looks good. FYI, the approach we've taken in flux-core is disable pylint warnings on a case-by-case basis using `pylint: disable=` comments inline, e.g.:
```python

    # pylint: disable=too-many-branches,too-many-statements
    def jobspec_create(self, args):
        """
        Create a jobspec from args and return it to caller
        """
```

I wonder if that is more appropriate here, rather than just disabling those checks everywhere? The result is you still get warnings for future pylint violations, and then can think about whether disabling the warnings make sense for each individual case.

**Comment by cmoussa1** [comment]
Oh! TIL! Didn't even know you could do this. I think that sounds like a fine approach. I'll try it out here and see if I can just add these disabled checks inline. Will report back if/when I have it working.

**Comment by cmoussa1** [comment]
I think that worked! Just force-pushed a fix to the last commit to add the disabled `pylint` checks inline instead of in the `.pylintrc` file.

File: src/cmd/flux-account-service.py
Comment (grondo) [reply src/cmd/flux-account-service.py]: Error message is better, but still a bit confusing:
```
ERROR:root:flux-accounting service is already registered
```
I don't know where the `root` is coming from?

Comment (cmoussa1) [reply src/cmd/flux-account-service.py]: Ah, agreed. The error message formatting is coming from the `logging` module, which is what I am using to emit the error. I _think_ by default, the module prints out the name of the logger, which in this case is `root`. But I can format the message to just not include the name, and instead only print out the type of error and the message, so that it will only output:

```console
ERROR:flux-accounting service is already registered
```

I can push that fix up right now.

File: src/cmd/flux-account-service.py
Comment (grondo) [inline src/cmd/flux-account-service.py:205]: `sqlite3.OperationalError` is a subclass of [sqlite3.Error](https://docs.python.org/3/library/sqlite3.html#sqlite3.Error), and does not appear to have an `errno` attribute.

I'd go through all of `flux-account-service.py` and check usage of exceptions to ensure attributes are correct. As noted above, also add some context to the error message, in case the string representation of the error doesn't indicate that it is a sqlite3 error.

Also, it might be good to consider how to test some of these error conditions, though I'm not sure how to do that.

**Review by grondo — COMMENTED** [review]
I made another pass focusing a bit more on error handling. See comments inline.

File: src/cmd/flux-account-service.py
Comment (grondo) [reply src/cmd/flux-account-service.py:152]: Currently you get this message, though admittedly the likelihood of this is very low:
```console
$ flux python -c 'import flux; print(flux.Flux().rpc("accounting.add_user", {}).get())'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/lib64/flux/python3.6/flux/rpc.py", line 77, in get
    resp_str = self.get_str()
  File "/usr/lib64/flux/python3.6/flux/util.py", line 90, in func_wrapper
    retval = func(future, *args, **kwargs)
  File "/usr/lib64/flux/python3.6/flux/rpc.py", line 71, in get_str
    self.pimpl.flux_rpc_get(payload_str)
  File "/usr/lib64/flux/python3.6/flux/util.py", line 66, in func_wrapper
    raise EnvironmentError(error.errno, errmsg.decode("utf-8")) from None
OSError: [Errno 22] 'username'

```

File: src/cmd/flux-account-service.py
Comment (cmoussa1) [reply src/cmd/flux-account-service.py:118]: Yeah, the behavior of the `view-user` command doesn't fail if the user can't be found in the flux-accounting DB, it just returns `User not found in association_table` if it can't find it. This has always been the behavior, but I can definitely open an issue on this and work on this on the side. 👍 I'll change it to instead return a non-zero exit code and perhaps log an error message that it could not find the user.

File: src/cmd/flux-account-service.py
Comment (cmoussa1) [reply src/cmd/flux-account-service.py:155]: Yeah, actually, after looking through the `try-except` blocks in a lot of these functions, it turns out the exceptions being listed here are already handled in the bindings themselves, so I'm not sure I need them to be listed here. I think I can just handle the `KeyError` exception and a general `Exception` for any non-`OSError` not already handled by the bindings. So maybe I should just remove the various `ValueError`, `sqlite3.OperationalError` and `IntegrityError`, etc. from `flux-account-service.py`.

Perhaps a near PR in the future should consist of providing more context and clarity to the exceptions being handled in the bindings themselves (e.g the example you provided in your comment: `f"sqllite3 operational error: {op_err}"`). I can open a separate issue on this.

Comment (grondo) [reply src/cmd/flux-account-service.py:155]: I know pylint throws an error if you use general `Exception` in a try/except block, but you must have that suppressed here.
But yes, if the error messages in the exception are already well formatted, then all you have to do is something like

```python
    except (ValueError, OSError, sqlite3.Error) as exc:
        handle.respond_error(msg, 0, str(exc))
```

Comment (grondo) [reply src/cmd/flux-account-service.py:155]: Yeah, that seems right, though it seems more "pythonic" to just let the ValueError be raised to the caller?

Comment (grondo) [reply src/cmd/flux-account-service.py:155]: I would open a separate issue functions that return strings instead of raising an exception on error. This is probably ok for now, but it seems like the caller should be able to tell the difference between success and failure for these functions?

Comment (cmoussa1) [reply src/cmd/flux-account-service.py:155]: Agreed. I can open a separate issue on converting the functions that return strings on error to instead raise the appropriate exception. That should probably be near-future work, if not work immediately following this PR. Does that sound ok?

Comment (grondo) [reply src/cmd/flux-account-service.py]: Yes, sounds good

Comment (cmoussa1) [reply src/cmd/flux-account-service.py]: Cool. I force-pushed up a change to remove the `except` blocks for `Exception`s currently handled by the bindings.

**Review by grondo — COMMENTED** [review]
A couple more items I noticed this time around.
Sorry I didn't catch these earlier!

File: src/cmd/flux-account-service.py
Comment (cmoussa1) [reply src/cmd/flux-account-service.py:115]: Oops! Yeah, while I was looking at this I noticed that I wasn't shutting down the service in the tests either, so I wouldn't have caught this. Thanks. I just force-pushed a fix to add a `handle.respond(signum)` to the `"shutdown"` message as well as a test in all of the sharness tests that use the service to call `"shutdown"` at the end of the test file.

File: src/cmd/flux-account.py
Comment (cmoussa1) [reply src/cmd/flux-account.py]: Ah, I didn't know this, thanks for pointing that out. I just force-pushed a fix to remove the `json.dumps()` from the RPCs in `flux-account.py`.

File: src/cmd/flux-account-service.py
Comment (cmoussa1) [reply src/cmd/flux-account-service.py:115]: Just realized that I don't think that `handle.respond(signum)` is the right way to respond to this message, as when I go to shut down the service, there is some error messages that pop up:

```console
flux-uri: ERROR: 
Invalid arguments passed to wrapped C function:
Name: flux_respond
C signature: int(*)(struct flux_handle_struct *, struct flux_msg *, char *)
Arguments: (15, <cdata 'void *' NULL>)
          

[1]+  Exit 1                  flux account-service -p /tmp/test.db
```

Do you know of the right way to respond to this message? The parameters for this function `(self, handle, watcher, signum, arg)` are different from the other functions `(self, handle, watcher, msg, arg)`.

Comment (grondo) [reply src/cmd/flux-account-service.py:115]: Ah, you are using the same callback for signal watcher and the `shutdown` message handler, so the arguments are not correct. You'll need a separate handler for the message watcher, with the same parameters as the other message watcher, then use `handle.respond(msg)` (which sends a response with an empty payload)

Comment (cmoussa1) [reply src/cmd/flux-account-service.py:115]: Oh okay! Thanks, I think that makes sense now. In this case, the signal watcher does not need to respond since it does not take a message from an RPC, correct? Its shutdown process can just end with `self.handle.reactor_stop()`?

Comment (cmoussa1) [reply src/cmd/flux-account-service.py:115]: Nonetheless, I just force-pushed a fix to correct the signal watcher and added a `shutdown_service` message handler (the latter actually responds with `handle.respond(msg)`.

**Review by grondo — APPROVED** [review]

**Review by grondo — APPROVED** [review]
Looks good with your latest changes @cmoussa1. Nice work!

Next step will be to add a systemd unit file and get that working. We can experiment using the fluxorama docker container

**Comment by cmoussa1** [comment]
@grondo thank you! This took a lot of iterations so I appreciate you reviewing and working through the revisions with me.

Do you think adding a systemd unit file should also be a component of this PR, or should it be proposed in a subsequent PR?

File: src/cmd/flux-account-service.py
Comment (grondo) [reply src/cmd/flux-account-service.py:115]: Yeah, the signal watcher has nothing to respond to, so no response needed there.

**Comment by grondo** [comment]
I think it could be a separate PR - so you can get this chunk of work merged and we can iterate on the systemd unit file (and how to test it)

**Comment by cmoussa1** [comment]
That sounds like a good plan to me! I'll go ahead and set MWP on this PR and then open an issue on adding a systemd unit file to flux-accounting - that could be work following this PR. Thanks!

**Pull request merged**
